core/trino-main/src/main/java/io/trino/cache/CanonicalSubplanExtractor.java

```diff
@@ -0,0 +1,629 @@
+                }
+                else if (!originalSymbol.equals(symbol)) {
+                    // aliasing of column id to multiple symbols is not supported
+                    if (extendSubplan) {
+                        canonicalSubplans.add(subplan);
+                    }
+                    return Optional.empty();
+                }
+            }
+
+            // Unsafe expressions that could throw an error should be evaluated only for the rows from
+            // original subquery. Therefore, common subplan predicate must match original subplan predicate.
+            // If common subplan predicate is wider, then unsafe expressions could fail even though
+            // evaluation of the original subplan would be successful.
+            boolean safeProjections = node.getAssignments().getExpressions().stream().noneMatch(expression -> mayFail(plannerContext, expression));
+            Set<Expression> requiredConjuncts = !safeProjections ? subplan.getPullableConjuncts() : ImmutableSet.of();
+
+            CanonicalSubplanBuilder builder = extendSubplan ?
+                    CanonicalSubplan.builderExtending(setRequiredConjuncts(subplan.getKey(), requiredConjuncts), subplan) :
+                    CanonicalSubplan.builderForChildSubplan(new FilterProjectKey(requiredConjuncts), subplan);
+            return Optional.of(builder
+                    .originalPlanNode(node)
+                    .originalSymbolMapping(symbolMappingBuilder.buildOrThrow())
+                    .assignments(assignments)
+                    // all symbols (and thus conjuncts) are pullable through projection
+                    .pullableConjuncts(subplan.getPullableConjuncts())
+                    .build());
+        }
+
+        private Key setRequiredConjuncts(Key key, Set<Expression> requiredConjuncts)
+        {
+            switch (key) {
+                case ScanFilterProjectKey scanFilterProjectKey -> {
+                    checkArgument(scanFilterProjectKey.requiredConjuncts().isEmpty());
+                    return new ScanFilterProjectKey(scanFilterProjectKey.tableId(), requiredConjuncts);
+                }
+                case FilterProjectKey filterProjectKey -> {
+                    checkArgument(filterProjectKey.requiredConjuncts().isEmpty());
+                    return new FilterProjectKey(requiredConjuncts);
+                }
+                default -> throw new IllegalStateException("Unsupported key type: " + key);
+            }
+        }
+
+        @Override
+        public Optional<CanonicalSubplan> visitFilter(FilterNode node, Void context)
+        {
+            PlanNode source = node.getSource();
+
+            if (containsLambdaExpression(node)) {
+                // lambda expressions are not supported
+                canonicalizeRecursively(source);
+                return Optional.empty();
+            }
+
+            if (!isDeterministic(node.getPredicate())) {
+                canonicalizeRecursively(source);
```

Comment:
We shouldn't call `canonicalizeRecursively` (here and in case of lambda expression) if the filter contains a dynamic filter, otherwise `DynamicFiltersChecker` might fail on `Dynamic filters [...] present in filter predicate whose source is not a table scan.`, for example:
```
@Test
public void testConjunctionOfNonDeterministicPredicateAndDynamicFilter()
{
    MaterializedResultWithPlan resultWithCache = executeWithPlan(
            withCacheEnabled(),
            """
                    SELECT l.partkey
                    FROM
                        (SELECT * FROM orders WHERE random(shippriority + 1) > 20) o
                    JOIN
                        (SELECT * FROM lineitem WHERE random(CAST(quantity AS INTEGER)) > 5) l
                    ON
                        l.ORDERKEY = o.ORDERKEY
                    """);
}
```

Reply:
Solution suggestion-
we can add the following method to `io.trino.sql.DynamicFilters`:
```
public static boolean containsDynamicFilter(Expression expression)
{
    return IrUtils.preOrder(expression)
            .anyMatch(DynamicFilters::isDynamicFilter);
}
```
and this one to `CanonicalSubplanExtractor`:
```
private void canonicalizeFilterSource(FilterNode node)
{
    if (!DynamicFilters.containsDynamicFilter(node.getPredicate())) {
        // filter's source must be a table scan if it contains a dynamic filter
        canonicalizeRecursively(node.getSource());
    }
}
```
and the test I've posted can make the following assertion:
```
// make sure only one side was cached (the one without dynamic filter)
assertThat(getOperatorStats(resultWithCache.queryId(), CacheDataOperator.class.getSimpleName()).count()).isEqualTo(1L);
```

Reply:
Indeed, it seems like an issue. we could probably probably extract parts of filters expressions event in that case, but your suggestions make sense

Reply:
@assaf2 instead of
```
public static boolean containsDynamicFilter(Expression expression)
{
    return IrUtils.preOrder(expression)
            .anyMatch(DynamicFilters::isDynamicFilter);
}
```
you could use:
```
        private boolean containsDynamicFilter(Expression expression)
        {
            return !extractDynamicFilters(expression).getDynamicConjuncts().isEmpty();
        }
```

Reply:
@assaf2 I've update PR.